src/AbstractNormalizedEventManager.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+namespace Dhii\WpEvents;
+
+use Psr\EventManager\EventInterface;
+
+/**
+ * Common functionality for WordPress event managers that use callback wrappers
+ * to normalize event handler parameters.
+ *
+ * @since [*next-version*]
+ */
+abstract class AbstractNormalizedEventManager extends AbstractWpEventManager
+{
+    /**
+     * Gets an event callback wrapper.
+     *
+     * @since [*next-version*]
+     * @see _createCallbackWrapper()
+     *
+     * @param string   $name     Name of the event.
+     * @param callable $callback Handler of the event.
+     *
+     * @return \Closure The wrapper.
+     */
+    protected function _getHandlerWrapper($name, $callback)
+    {
+        return $this->_createHandlerWrapper($name, $callback);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    protected function _attach($name, $callback, $priority = self::DEFAULT_PRIORITY, $numArgs = 1)
+    {
+        $handler = $this->_getHandlerWrapper($name, $callback);
+        $this->_addHook($name, $handler, $priority, $numArgs);
+
+        return $this;
+    }
+
+    /**
+     * Creates a wrapper for an event handler, containing and replacing it.
+     *
+     * If the wrapper receives an {@see EventInterface}, it will use that and assume it was
+     * triggered or normalized earlier by the event manager.
+     * Otherwise, it will assume that regular parameters are passed, such as
+     * what WordPress does, and will create an event object from that.
+     * Also allows events to stop propagation.
+     *
+     * @since [*next-version*]
+     *
+     * @param string   $name     The name of the event, for which a wrapper is created.
+     * @param callable $callback The callback to wrap.
+     *
+     * @return \Closure The wrapper.
+     */
+    protected function _createHandlerWrapper($name, $callback)
+    {
+        $me = $this;
+
+        return function () use ($name, &$callback, &$me) {
+            $args  = \func_get_args();
+            $event = \count($args) && $args[0] instanceof EventInterface
+                    ? $args[0]
+                    : $me->createEvent($name, $args);
+            /* @var $event \Dhii\WpEvents\Event */
+            if (!$event->isPropagationStopped()) {
+                \call_user_func_array($callback, array($event));
+            }
+
+            return $event;
```

Comment:
## Summary of Voice Call

The combined solution involves using a cache of `EventInterface` instances, mapped by the event name. Each event will not store its corresponding `EventInterface` instance in the cache for the duration of the event chain.

These cached instances will be created on demand: when a handler _wrapper_ attempts to get the event instance, the cache will create it if it doesn't exist and will also register a handler with priority `PHP_INT_MAX`.

This (semi-)ensures that the handler will be called last in the chain. It's purpose is simple: it will delete the event instance from the cache and unhook itself. This is so that if an event with the same name is later triggered, it receives a new event instance. In these cases, a new late handler will be registered with priority `PHP_INT_MAX` that will do the same thing.

All of this is to ensure that every chain of events gets the its own `EventInterface` instance that persists until the chain ends.

----

There was a bug in the [`AbstractWpEventManager#_detach`][1] where the priority argument was not being passed to `remove_filter`. I will be pushing a fix for this shortly.

[1]: https://github.com/Dhii/wp-events/blob/task/bring-up-to-dhii-standard_6/src/AbstractWpEventManager.php#L102